### PR TITLE
Remove MapView definitions from @types/react-native - create new react-native-maps module

### PR DESCRIPTION
**Pull request by methodbox**
initialRegion is a valid prop in `react-native-maps` but it's definition is missing in the @types.

This contains the same properties as the `region` and is optional, and therefore should be defined as 

```js
export interface MapViewProps extends ViewProps {
...
initialRegion?: MapViewRegion;
...
}
```
Which is the same as `region`.

Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
You can test this simply by placing the following code in a `.tsx` file and note that `MapView` is flagged for the `initialRegion` with no definition:
```js
import React, { Component } from 'react';
import { MapView } from 'expo';  //  or react-native-maps

export default class MapTest extends Component {
  render() {
    return (
      <MapView
        style={{ flex: 1 }}
        initialRegion={{
          latitude: 37.78825,
          longitude: -122.4324,
          latitudeDelta: 0.0922,
          longitudeDelta: 0.0421,
        }}
      />
    );
  }
}
```
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: 
This is defined in their component API docs here:
https://github.com/react-native-community/react-native-maps/blob/master/docs/mapview.md

### CODE REVIEW

**Comment by alloy** [comment]
Just to be clear, this prop exists when importing from `react-native` (the deprecated version)?

**Comment by methodbox** [comment]
> Just to be clear, this prop exists when importing from `react-native` (the deprecated version)?

No, sorry that was a typo.

https://github.com/react-native-community/react-native-maps/blob/master/lib/components/MapView.js

It exists in `react-native-maps` currently.  It's not part of the `react-native` core, but it's already in the `@types/react-native` definitions.

Not sure how exactly that should be handled.

Maybe it would be best to separate this into its own definitions file?

**Comment by alloy** [comment]
Ah right, yes then that needs to be separated out to its own package 👍 Feel free to copy the maintainer list in the current type defs. Thanks!

**Comment by methodbox** [comment]
> Ah right, yes then that needs to be separated out to its own package 👍 Feel free to copy the maintainer list in the current type defs. Thanks!

Ok, well it looks like the build in Travis failed again because of some mysterious white space I have in my edit.

Do I need a separate PR for this, or should I create the new types file under this one?

**Comment by alloy** [comment]
I don’t know if there are any rules around that, sorry :/ But feel free to test that out [and ping me on it]!

**Comment by methodbox** [comment]
@alloy

Okay, I think a separate package is most
appropriate, since the `MapView` component isn’t actually part of `react-native` and is now separated into the `react-native-maps` module, which is itself part of the React Native Community.

I’m going to try creating a separate package in `...types/react-native-maps`, I’ll rip the definitions from `react-native` into this and add the missing types included by the `react-native-maps` team along with my fix for the missing prop.

**Comment by methodbox** [comment]
As this module was deprecated by FB some time ago and is no longer supported I will be resubmitting this PR to remove the type definitions from @types/react-native.

This will solve the current problem of conflicts with the `react-native-maps` module https://github.com/react-native-community/react-native-maps as there are conflicting namespaces for `MapView` and it's related components, etc. which don't share the proper type definitions in @types/react-native. For now, this will allow that package to function correctly as it's `index.d.ts` seems to have the appropriate types for their version of `MapView`.

I'll submit a separate PR for creating an @types/react-native-maps module once they agree to my proposal for removing the `index.d.ts` and opting for using `react-native`'s own recommended method of using @types:  https://facebook.github.io/react-native/blog/2018/05/07/using-typescript-with-react-native - this is less likely to cause conflicts then them maintaining an index.d.ts in their module instead of on @types, as @types modules can be checked against the rest of @types for conflicts (which would have prevented this issue).

https://github.com/react-native-community/react-native-maps/issues/2847

**Comment by methodbox** [comment]
Reverted previous two commits. These will need to wait until the conflict here is resolved: https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/styled-components/native.d.ts

I will put in a separate PR for this.

**Comment by alloy** [comment]
Thanks for doing this, I agree with your approach 👍 

Just to be clear, currently your PR is a no-op, but the goal is to have this PR remove the MapView related types from the core RN one, right?

**Comment by jessetrinity** [comment]
@methodbox are you closing this PR and opening a separate one?

**Comment by methodbox** [comment]
@jessetrinity No I’m waiting for the PR here to be completed:  

https://github.com/DefinitelyTyped/DefinitelyTyped/pull/35274

I’ll need to resubmit the changes from my reverted commits after that happens. The `styled-components` is included on the Travis test so I wasn’t able to pass the checks to get this into a mergable status.

It still had some types that included `MapView` from `react-native`.

**Comment by methodbox** [comment]
> Thanks for doing this, I agree with your approach 👍
> 
> Just to be clear, currently your PR is a no-op, but the goal is to have this PR remove the MapView related types from the core RN one, right?

I should probably clarify - there _isn't a need_ to open a separate PR, right?

**Comment by alloy** [comment]
> there isn't a need to open a separate PR, right?

Nope, all commits will get squashed.

**Comment by methodbox** [comment]
Update: the `styled-components` related PR is about to get merged. Once it does, I’ll try this one again.

https://github.com/DefinitelyTyped/DefinitelyTyped/pull/35274#issuecomment-491539909

**Comment by methodbox** [comment]
Is it necessary for me to merge the **current master** with my master first? I know there have been other commits merged since I made my fork.

**Comment by sandersn** [comment]
@methodbox, no, although travis-ci/push will end up testing any changes that have been made to current master since you forked from master.

Is this ready to go in? Or do you still need to re-add the commit that adds react-native-maps?

**Comment by methodbox** [comment]
I'll still need to add react-native-maps. I'll try to work on this later today.

**Comment by alloy** [comment]
Let’s do it together, that way we know users of the new RN types will actually have the map types package available.

**Comment by methodbox** [comment]
_Just a heads up on incoming commits:_

I'm going to be pushing a number of commits shortly for the `react-native-maps` types module. I've written a couple of simple tests which are not complete, but will do so for each primary component.

Here's a list of the components, from their docs, first two are done:
```
<MapView />
<Marker />
<Callout />
<Polygon />
<Polyline />
<Circle />
<Overlay />
```

For now, I'm not writing tests for the events or methods because I think it's overkill as they primarily share the core types; these might show something on edge cases but as these types have been in use already for a bit I'm going to trust them for now. The tests will use an example of each of the props for the respective component.

As of the moment I have the types moved in and tests written for the `MapView` and `Marker`.

I'll comment when I feel it's ready to merge.

**Comment by methodbox** [comment]
Can someone help me make sense of this Travis error?

```bash
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! definitely-typed@0.0.3 test: `node node_modules/types-publisher/bin/tester/test.js --run-from-definitely-typed`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the definitely-typed@0.0.3 test script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
npm ERR! A complete log of this run can be found in:
npm ERR!     /home/travis/.npm/_logs/2019-05-16T01_46_02_683Z-debug.log
The command "npm run test" exited with 1.
```

**Comment by methodbox** [comment]
Okay, I'm lost. I have no idea why it's failing. I pushed a new tslint.json because I left this out, but I don't understand why it's failing.

File: types/react-native-maps/index.d.ts [hunk 1]
```diff
@@ -0,0 +1,474 @@
+declare module "react-native-maps" {
```
Comment (alloy) [inline types/react-native-maps/index.d.ts]: This file should start with headers that specify what this package is, its version, the maintainers, etc. You can see that CI is failing at the ‘parse header’ stage https://travis-ci.org/DefinitelyTyped/DefinitelyTyped/builds/533108975#L477

Copy it from the RN types and adjust the name and version number.

Comment (alloy) [reply types/react-native-maps/index.d.ts]: Btw, there’s no need for `declare module "react-native-maps" { … }` when you have a `.d.ts` file that maps directly to the import of the module.

Comment (methodbox) [reply types/react-native-maps/index.d.ts]: Ok awesome. Thanks for the explanation!

I’ll take a look later today.

File: types/react-native-maps/index.d.ts [hunk 2]
```diff
@@ -0,0 +1,522 @@
...
+//                 Ho Yin (Michael) Cheng <https://github.com/hysan>
+//                 Guilherme Eduardo Gulinoski Teixeira <https://github.com/GuilhermeEd>
+//                 lantz moore <https://github.com/mizerlou>
+//                 Almir Kadric <https://github.com/AlmirKadric>
+//                 origins-hongyi <https://github.com/fozhao>
+//                 Christoph <https://github.com/Stophface>
+//                 Budi Adiono <https://github.com/budiadiono>
+//                 Alexander Veligurov <https://github.com/alvelig>
+//                 Adrian Stutz <https://github.com/sttz>
+//                 Chris Shaffer <https://github.com/methodbox>
```
Comment (alloy) [inline types/react-native-maps/index.d.ts]: Why is this list so much longer than the one of RN? https://github.com/DefinitelyTyped/DefinitelyTyped/blob/0d753f7c4c38f158613c7b0d22a554900a09afc7/types/react-native/index.d.ts#L3-L23

Comment (methodbox) [reply types/react-native-maps/index.d.ts]: Because the last one is me ;) and the rest are from the actual types contributors from `react-native-maps` types - their `index.d.ts`.

I’ve proposed to them already moving this to @types; before they process my PR with them to remove their `index.d.ts` we are waiting for this to be completed.

However, I included them because those are ultimately the people who contributed the types, with the exception of one that I corrected.

Comment (alloy) [reply types/react-native-maps/index.d.ts]: Totally, but do they realise that this would mean they will start receiving notifications for each issue/PR about these types on this repo?

Comment (methodbox) [reply types/react-native-maps/index.d.ts]: Oh! I did not realize that, either. I can remove that then if that’s the best move?

Comment (methodbox) [reply types/react-native-maps/index.d.ts]: Well, I've removed them. I guess they can be added later if necessary.

**Comment by methodbox** [comment]
...this travis config is VERY picky .. can't even have a `.gitignore`??  Hopefully 8th time is a charm ;).

**Comment by methodbox** [comment]
I think this is ready to merge; I'm going to leave tests out of this for now as these types have already been in use in the wild already for some time, and the build is passing it is.

Obviously you don't need my okay for that ;) but I'm just confirming that I'm not pushing anything else to this PR.

File: types/react-native-maps/index.d.ts (hunk 2 above)
Comment (alloy) [reply types/react-native-maps/index.d.ts]: I always struggle with this header. The name makes it seem like it’s just an acknowledgement of who contributed, not that it has the side-effect of signing up as maintainer 🤷‍♂ So yeah, I would err on the side of caution too. Perhaps let them know on the upstream PR that they can sign-up as maintainer here.

**Comment by alloy** [comment]
Yeah I’m ok with that, but I _would_ urge you to start adding coverage in new additions (either yours or when you review other’s additions).

**Comment by alloy** [comment]
Thanks! Great work on seeing this through! 🙏

Let’s do this 🎉

**Comment by methodbox** [comment]
Already done. RNM has been informed and I’ve let them know I can or they can do an PR to add themselves, etc. and I’ve linked back here.

On May 19, 2019, at 2:26 PM, Eloy Durán <notifications@github.com<mailto:notifications@github.com>> wrote:


@alloy commented on this pull request.

________________________________

In types/react-native-maps/index.d.ts<https://github.com/DefinitelyTyped/DefinitelyTyped/pull/35239#discussion_r285394076>:

> +//                 Mellet Solbakk <https://github.com/mellet>
+//                 Mellet Solbakk <https://github.com/mellet>
+//                 Nizam Moidu <https://github.com/netmaxt3r>
+//                 Marcel Kalveram <https://github.com/marcelkalveram>
+//                 Jason Li <https://github.com/lijason103>
+//                 JulienR2 <https://github.com/julienR2>
+//                 Ho Yin (Michael) Cheng <https://github.com/hysan>
+//                 Guilherme Eduardo Gulinoski Teixeira <https://github.com/GuilhermeEd>
+//                 lantz moore <https://github.com/mizerlou>
+//                 Almir Kadric <https://github.com/AlmirKadric>
+//                 origins-hongyi <https://github.com/fozhao>
+//                 Christoph <https://github.com/Stophface>
+//                 Budi Adiono <https://github.com/budiadiono>
+//                 Alexander Veligurov <https://github.com/alvelig>
+//                 Adrian Stutz <https://github.com/sttz>
+//                 Chris Shaffer <https://github.com/methodbox>


I always struggle with this header. The name makes it seem like it’s just an acknowledgement of who contributed, not that it has the side-effect of signing up as maintainer 🤷‍♂ So yeah, I would err on the side of caution too. Perhaps let them know on the upstream PR that they can sign-up as maintainer here.
